Deprecate MapChange and provide dedicated callback methods

tobrun opened the pull request:
This PR deprecates using OnMapChangeListeners and replaces them with dedicated callbacks.
 - [x] jni bindings
 - [x] public api for dedicated callbacks
 - [x] javadoc on public API
 - [x] refactor internal use of MapChange with dedicated callbacks 
 - [x] deprecate old setup (OnMapChangeListener) 
 - [x] optimise `onMapFailedLoaded` with the thrown error message
 - [x] optimise `onSourceChanged` callback to provide source/attribution that has changed
 - [x] optimise "partial" vs "full" callbacks (expose same as core in terms of amount of events)
 - [x] add tests
 - [x] update test app

Proposed API:

```java
    void onCameraWillChange(boolean animated);
    void onCameraIsChanging();
    void onCameraDidChange(boolean animated);

    void onWillStartLoadingMap();
    void onDidFinishLoadingMap();
    void onDidFailLoadingMap(String errorMessage);

    void onWillStartRenderingFrame();
    void onDidFinishRenderingFrame(boolean partial);

    void onWillStartRenderingMap();
    void onDidFinishRenderingMap(boolean partial);

    void onDidFinishLoadingStyle();

    void onSourceChangedListener(String id);
```

Closes #8389 / refs #9476

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapboxMap.java:
```diff
@@ -184,21 +184,32 @@ void onPostMapReady() {
   /**
    * Called when the region is changing or has changed.
```
Is this javadoc still correct?

ivovandongen commented on platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/MapView.java:
```diff
@@ -174,6 +175,9 @@ private void initialise(@NonNull final Context context, @NonNull final MapboxMap
     nativeMapView.setReachability(ConnectivityReceiver.instance(context).isConnected(context));
 
+    // bind internal components for map change events
+    mapChangeDispatcher.bind(mapCallback = new MapChangeResultHandler(mapboxMap));
```
Would it somehow be feasible to make this a constructor argument? Since it is required. That would make it possible to skip a lot of null checks in `MapChangeDispatcher`

tobrun replied:
Been looking into it but don't think so at the moment though this clearly shows we need to rethink our architecture. The issue here is a cyclic class dependency: NativeMapview depends on MapChangeDispatcher, MapboxMap depends on NativeMapView and MapChangeDispatcher depends on MapboxMap. 

Somewhere during initialization we need to bind them together and require us to have null checking. Current setup follows the requirement of being able to listen to map change events as soon as possible. I don't see an a issue with internal components missing out on the initialization change events.

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -54,14 +54,14 @@
 namespace mbgl {
 namespace android {
 
-NativeMapView::NativeMapView(jni::JNIEnv& _env,
```
Could you revert indentation changes?

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -54,14 +54,14 @@
-    : javaPeer(_obj.NewWeakGlobalRef(_env)),
-      pixelRatio(_pixelRatio),
-      threadPool(sharedThreadPool()) {
+        : javaPeer(_obj.NewWeakGlobalRef(_env)),
```
dito 3x

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -112,7 +112,7 @@ void NativeMapView::bind() {
  * From mbgl::RendererBackend.
  */
-gl::ProcAddress NativeMapView::initializeExtension(const char* name) {
+gl::ProcAddress NativeMapView::initializeExtension(const char *name) {
```
dito

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -171,110 +171,183 @@ void NativeMapView::deactivate() {
  */
 void NativeMapView::invalidate() {
     android::UniqueEnv _env = android::AttachEnv();
-    static auto onInvalidate = javaClass.GetMethod<void ()>(*_env, "onInvalidate");
```
dito

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -171,110 +171,183 @@ void NativeMapView::deactivate() {
  *
  * May be called from any thread
  */
-void NativeMapView::notifyMapChange(mbgl::MapChange change) {
```
Please also remove the declaration here: https://github.com/mapbox/mapbox-gl-native/blob/master/platform/android/src/native_map_view.hpp#L67

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -171,110 +171,183 @@ void NativeMapView::deactivate() {
+void NativeMapView::onDidFailLoadingMap(std::exception_ptr eptr) {
+    assert(vm != nullptr);
+    try {
+        std::rethrow_exception(eptr);
```
Do we want to throw an exception here? Seems like we always want to invoke the listener. If re-throwing is required, it should happen where the exception is produced.

tobrun replied:
rethrowing was a way of getting the string message for the listener, let me try a different approach.

tobrun replied:
the exception is produced in core in `style_impl.cpp` and passed to the MapObserver, which invokes the method above. Don't think I can move re-throwing up the chain as it would impact other bindings. The code for retrieving the exception message was based of http://en.cppreference.com/w/cpp/error/exception_ptr.

ivovandongen replied:
No, you're right. That's a valid way to do it. Should've read more carefully.

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -171,110 +171,183 @@ void NativeMapView::deactivate() {
 
 /**
- * From mbgl::RendererBackend. Callback to java NativeMapView#onMapChanged(int).
+ * From mbgl::RendererBackend. Callback to java NativeMapView#onCameraWillChange().
```
Actually. This is from `MapObserver` and will only be coming from the main (map) thread. Could you update the comment in the header file as well?

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -171,110 +171,183 @@ void NativeMapView::deactivate() {
 void NativeMapView::onCameraIsChanging() {
-    notifyMapChange(MapChange::MapChangeRegionIsChanging);
+    assert(vm != nullptr);
+    android::UniqueEnv _env = android::AttachEnv();
```
As this is not actually coming from any other thread then the main/map thread, this is not needed.

tobrun replied:
I changed it to something as:

```
void NativeMapView::onDidFinishLoadingStyle() {
    JNIEnv* env = nullptr;
    int resultCode = vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6);
    if(resultCode == JNI_OK) {
        static auto onDidFinishLoadingStyle = javaClass.GetMethod<void()>(*env, "onDidFinishLoadingStyle");
        javaPeer->Call(*env, onDidFinishLoadingStyle);
    }
}
```

lmk if there is a better way of getting the `env`

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -171,110 +171,183 @@ void NativeMapView::deactivate() {
     _createSurface(ANativeWindow_fromSurface(&env, jni::Unwrap(*_surface)));
 }
 
-void NativeMapView::destroySurface(jni::JNIEnv&) {
```
more formatting changes

tobrun commented:
I redid this PR in light of moving to GLSurfaceView (reason this PR was on hold). 
A couple of changes since previous iteration:
 - we allow multiple listeners to be registered (this was required for plugin development, else a user or a plugin could override an already set listener).
 - usage of AttachEnv to get a reference to the jni environment

ivovandongen commented on platform/android/src/native_map_view.cpp:
```diff
@@ -94,80 +94,142 @@ NativeMapView::~NativeMapView() {
+void NativeMapView::onDidFailLoadingMap(std::exception_ptr eptr) {
+    try {
+       if(eptr) {
+           std::rethrow_exception(eptr);
```
You can use https://github.com/mapbox/mapbox-gl-native/blob/master/include/mbgl/util/string.hpp#L52 instead.

tobrun commented:
Thank you for the review @ivovandongen, 
re using toString for the exception ptr, this has been fixed.
re naming: I merged the concepts of `MapChangeDispatcher` and `MapChangeHandler` into `MapChangeEventManager`, feel free propose any changes to the name.

tobrun commented:
stale
